openpgp/ocb/ocb.go

```diff
@@ -0,0 +1,269 @@
+	X := make([]byte, len(text))
+	copy(X, text)
+	blockSize := o.block.BlockSize()
+
+
+	//
+	// Nonce-dependent and per-encryption variables
+	//
+	// From RFC 7253:
+	// Nonce = num2str(TAGLEN mod 128, 7) || zeros(120 - bitlen(N)) || 1 || N
+	paddedNonce := append(make([]byte, blockSize-1-len(nonce)), 1)
+	paddedNonce = append(paddedNonce, nonce...)
+	paddedNonce[0] = byte(8*o.tagSize % blockSize)
+
+	bottom := int(paddedNonce[blockSize-1] & 63)
+	Ktop := make([]byte, blockSize)
+	copy(Ktop, paddedNonce)
+	// Zero out last 6 bits of Ktop
+	Ktop[blockSize-1] &= 192
+	o.block.Encrypt(Ktop, Ktop)
+
+	// Stretch = Ktop || ((lower half of Ktop) XOR (lower half of Ktop << 8))
+	xorHalves := bytesutils.XorBytes(Ktop[:blockSize/2], Ktop[1:1+blockSize/2])
+	stretch := append(Ktop, xorHalves...)
+	offset := make([]byte, len(stretch))
+	copy(offset, stretch)
+	offset = bytesutils.ShiftNBytesLeft(offset, bottom)
+	offset = offset[:blockSize]
+	checksum := make([]byte, blockSize)
+
+	//
+	// Process any whole blocks
+	//
+	// Note: For encryption Y is ciphertext || tag, for decryption Y is
+	// plaintext || tag.
+	Y := make([]byte, 0)
+	m := len(text) / blockSize
+	for i := 0; i < m; i++ {
+		index := bits.TrailingZeros(uint(i+1))
+		if len(o.mask.L)-1 < index {
+			o.mask.extendTable(index)
+		}
+		bytesutils.XorBytesMut(offset, o.mask.L[bits.TrailingZeros(uint(i+1))])
+		blockX := X[i*blockSize : (i+1)*blockSize]
+		blockY := bytesutils.XorBytes(blockX, offset)
+		switch instruction {
+		case "enc":
+			o.block.Encrypt(blockY, blockY)
+			bytesutils.XorBytesMut(blockY, offset)
+			bytesutils.XorBytesMut(checksum, blockX)
+		case "dec":
+			o.block.Decrypt(blockY, blockY)
+			bytesutils.XorBytesMut(blockY, offset)
+			bytesutils.XorBytesMut(checksum, blockY)
+		}
+		Y = append(Y, blockY...)
+	}
+
+	//
+	// Process any final partial block and compute raw tag
+	//
+	if len(X) % blockSize != 0 {
+		bytesutils.XorBytesMut(offset, o.mask.L_ast)
+		pad := make([]byte, blockSize)
+		o.block.Encrypt(pad, offset)
+		chunkX := X[blockSize*m:]
+		chunkY := bytesutils.XorBytes(chunkX, pad[:len(chunkX)])
+		// P_* || bit(1) || zeroes(127) - len(P_*)
+		switch instruction {
+		case "enc":
+			paddedY := append(chunkX, 128)
+			paddedY = append(paddedY, make([]byte, blockSize-len(chunkX)-1)...)
+			bytesutils.XorBytesMut(checksum, paddedY)
+		case "dec":
+			paddedX := append(chunkY, 128)
+			paddedX = append(paddedX, make([]byte, blockSize-len(chunkY)-1)...)
+			bytesutils.XorBytesMut(checksum, paddedX)
+		}
+		tag := bytesutils.XorBytes(checksum, offset)
+		bytesutils.XorBytesMut(tag, o.mask.L_dol)
+		o.block.Encrypt(tag, tag)
+		bytesutils.XorBytesMut(tag, o.hash(adata))
+		Y = append(Y, chunkY...)
+		Y = append(Y, tag[:o.tagSize]...)
+	} else {
+		tag := bytesutils.XorBytes(checksum, offset)
+		o.block.Encrypt(tag, bytesutils.XorBytes(tag, o.mask.L_dol))
+		bytesutils.XorBytesMut(tag, o.hash(adata))
+		Y = append(Y, tag[:o.tagSize]...)
+	}
+	return Y
+}
+
```

Comment:
I should add a comment on this hash function (the design property that `hash([])=` a sequence of zeroes)